xlive/Blam/Engine/game/players.cpp

```diff
@@ -306,11 +311,58 @@ void __cdecl player_validate_configuration(datum player_index, s_player_properti
     return;
 }
 
+// Basic void* type because actually "usercall"
+void* p_player_representation_get;
+
+// Calls the original player representation get function
+void player_representation_get_orig_fn(int player_index, int* out_variant_index, int* a3)
+{
+    __asm {
+        mov eax, a3
+        push eax
+        mov ebx, out_variant_index
+        mov eax, player_index
+        call p_player_representation_get
+        add esp, 4
+    }
+}
+
+void __cdecl player_representation_get(datum player_datum, int* out_variant_index, int* a3)
+{
+    s_player* player = s_player::GetPlayer(DATUM_INDEX_TO_ABSOLUTE_INDEX(player_datum));
+
+    player_configuration_validate_character_type(&player->properties[0]);
+
+    // call the original function after validation
+    player_representation_get_orig_fn(player_datum, out_variant_index, a3);
+}
+
+__declspec(naked) void player_representation_get_to_cdecl()
+{
+    __asm {
+        push eax
+        mov eax, [esp + 4h + 4h]
+        push eax // a3
+        push ebx // out_variant
+        mov eax, [esp + 8h]
+        push eax // player index
+        call player_representation_get
+        add esp, 10h // clear 16 bytes to esp == ret addr
+        retn
+    }
+}
+
+
 void players_apply_patches(void)
 {
     // Change the validation for player_appearance_valid to use the updated k_player_character_type_count constant
     WriteValue<BYTE>(Memory::GetAddress(0x54fb3, 0x5D4AB), k_player_character_type_count);
 
     // Replace the player profile validation function with our own
     PatchCall(Memory::GetAddress(0x5509E, 0x5D596), player_validate_configuration);
+
+    // Validate the player character type pre-spawn
+    p_player_representation_get = Memory::GetAddress<void*>(0x53895, 0x5BD8D);
+    PatchCall(Memory::GetAddress(0x559F9, 0x5DEF1), player_representation_get_to_cdecl);
+    PatchCall(Memory::GetAddress(0x53969, 0x5BE61), player_representation_get_to_cdecl);
```

Comment:
These functions shouldn't be hooked for verification.
Detour the player_spawn function instead and add a call to player_configuration_validate_character_type.

Reply:
uh, this is what this does? it hooks a function within player spawn to validate the character type pre-spawn

Reply:
I meant detouring player_spawn directly and executing the validation before player_spawn.
The function you're hooking currently is used twice during one of the spawn processes which is unnecessary.

Reply:
I don't see a problem with it happening twice